src/plugins/mavlink_ftp/mavlink_ftp_impl.cpp

```diff
@@ -0,0 +1,971 @@
+#include <functional>
+#include <iostream>
+#include <experimental/filesystem>
+
+#include "crc32.h"
+#include "mavlink_ftp_impl.h"
+#include "system.h"
+#include "global_include.h"
+
+namespace mavsdk {
+
+using namespace std::placeholders; // for `_1`
+
+namespace fs = std::experimental::filesystem;
+
+MavlinkFTPImpl::MavlinkFTPImpl(System &system) 
+    : PluginImplBase(system)
+{
+    _parent->register_plugin(this);
+}
+
+MavlinkFTPImpl::~MavlinkFTPImpl()
+{
+    _parent->unregister_plugin(this);
+}
+
+void MavlinkFTPImpl::init()
+{
+    _parent->register_mavlink_message_handler(
+        MAVLINK_MSG_ID_FILE_TRANSFER_PROTOCOL,
+        std::bind(&MavlinkFTPImpl::process_mavlink_ftp_message, this, _1),
+        this);
+}
+
+void MavlinkFTPImpl::deinit()
+{
+}
+
+void MavlinkFTPImpl::enable()
+{
+}
+
+void MavlinkFTPImpl::disable()
+{    
+}
+
+void MavlinkFTPImpl::_process_ack(PayloadHeader *payload)
+{
```

Comment:
I think you'll need a mutex around all the member vars, right?

Reply:
I implemented and tested it along with retry / timeout mechanism.